### PR TITLE
FRAME: Assets Freezer pallet

### PR DESCRIPTION
**Pull request by gavofyork**
Probably pretty stale now, but...

Polkadot companion: https://github.com/paritytech/polkadot/pull/3131

TODO: 
- [ ] Rename `AssetsFreezer` to `AssetsHolder`.
- [ ] Tweak `AssetsHolder` to implement the new `fungibles::hold::*` API.
- [ ] Introduce new `AssetsFreezer` which implements `fungibles::freeze::*` API.
- [ ] Tests.

### CODE REVIEW

File: frame/assets-freezer/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		/// The fungibles trait impl whose assets this reserves.
+		type Assets: fungibles::Inspect<Self::AccountId>;
+
+		/// Place to store the fast-access freeze data for the given asset/account.
+		type Store: StoredMap<(AssetIdOf<Self>, Self::AccountId), FreezeData<BalanceOf<Self>>>;
+
+//		/// Weight information for extrinsics in this pallet.
+//		type WeightInfo: WeightInfo;
```
Comment (apopiak) [inline frame/assets-freezer/src/lib.rs:91]: missing weights, not ready for review?

Comment (gavofyork) [reply frame/assets-freezer/src/lib.rs:91]: Not yet; for some reason the bot saw fit to remove my `inprogress` label and introduce `pleasereview`.

**Comment by xlc** [comment]
Holds should be named like https://github.com/paritytech/substrate/pull/7778

**Review by kianenigma — COMMENTED** [review]
@riusricardo what's the status quo here?

**Review by riusricardo — COMMENTED** [review]
@kianenigma a review would be good now.

File: frame/assets/src/lib.rs [hunk 2]
```diff
@@ -421,7 +425,7 @@ pub mod pallet {
 		///
 		/// Weight: `O(1)`
 		#[pallet::weight(T::WeightInfo::force_create())]
-		pub(super) fn force_create(
+		pub fn force_create(
```
Comment (riusricardo) [inline frame/assets/src/lib.rs:428]: @kianenigma Just want to highlight this changes as discussed.

File: frame/assets/src/lib.rs [hunk 3]
```diff
@@ -525,7 +529,7 @@ pub mod pallet {
 		/// Weight: `O(1)`
 		/// Modes: Pre-existing balance of `beneficiary`; Account pre-existence of `beneficiary`.
 		#[pallet::weight(T::WeightInfo::mint())]
-		pub(super) fn mint(
+		pub fn mint(
```
Comment (riusricardo) [inline frame/assets/src/lib.rs:532]: @kianenigma Just want to highlight this changes as discussed.

File: frame/assets-freezer/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+[dependencies]
+serde = { version = "1.0.101", optional = true }
+codec = { package = "parity-scale-codec", version = "2.0.0", default-features = false }
+sp-std = { version = "3.0.0", default-features = false, path = "../../primitives/std" }
+# Needed for various traits. In our case, `OnFinalize`.
+sp-runtime = { version = "3.0.0", default-features = false, path = "../../primitives/runtime" }
+# Needed for type-safe access to storage DB.
+frame-support = { version = "3.0.0", default-features = false, path = "../support" }
+# `system` module provides us with all sorts of useful stuff and macros depend on it being around.
```
Comment (kianenigma) [inline frame/assets-freezer/Cargo.toml:23]:
```suggestion
sp-runtime = { version = "3.0.0", default-features = false, path = "../../primitives/runtime" }
frame-support = { version = "3.0.0", default-features = false, path = "../support" }
```

reminiscence of `pallet-example` everywhere :D

File: frame/assets-freezer/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,361 @@
...
+
+	/// The information concerning our freezing.
+	#[derive(Eq, PartialEq, Clone, Encode, Decode, RuntimeDebug, Default)]
+	pub struct FreezeData<Balance> {
+		/// The amount of funds that have been reserved. The actual amount of funds held in reserve
+		/// (and thus guaranteed of being unreserved) is this amount less `melted`.
+		///
+		/// If this `is_zero`, then the account may be deleted. If it is non-zero, then the assets
+		/// pallet will attempt to keep the account alive by retaining the `minimum_balance` *plus*
+		/// this number of funds in it.
```
Comment (kianenigma) [inline frame/assets-freezer/src/lib.rs:60]: where is this logic enforced?

Comment (kianenigma) [reply frame/assets-freezer/src/lib.rs:60]: Okay, via `FrozenBalance` spec. Might be worth documenting it.

File: frame/assets-freezer/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
+}
+
+impl<T: Config> FrozenBalance<AssetIdOf<T>, T::AccountId, BalanceOf<T>> for Pallet<T> {
+	fn frozen_balance(id: AssetIdOf<T>, who: &T::AccountId) -> Option<BalanceOf<T>> {
+		let f = T::Store::get(&(id, who.clone()));
+		if f.reserved.is_zero() { None } else { Some(f.reserved) }
+	}
+}
+
+impl<T: Config> fungibles::Inspect<<T as SystemConfig>::AccountId> for Pallet<T> {
```
Comment (kianenigma) [inline frame/assets-freezer/src/lib.rs:114]: DQ: why does this pallet itself need to implement this? Anywhere where this is used, a `T::Assets` is also present where it does implement `Fungibles`. Internally, `T::Assets::fn` and externally `Runtime::AssetFreezer::Assets::fn` can expose everything needed.

Comment (kianenigma) [reply frame/assets-freezer/src/lib.rs:114]: ahh `T::Assets` is only `InspectWithoutFreezer`.

File: frame/assets/src/lib.rs (hunk 2 above)
Comment (kianenigma) [reply frame/assets/src/lib.rs:428]: Looking at the PR a bit more, can't you use `T::Assets` interface of the AssetFreezer to do this?

File: frame/assets-freezer/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,361 @@
...
+impl<T: Config> fungibles::MutateHold<<T as SystemConfig>::AccountId> for Pallet<T> where
+	T::Assets: fungibles::Transfer<T::AccountId> + fungibles::InspectWithoutFreezer<T::AccountId>
+{
+	fn hold(asset: AssetIdOf<T>, who: &T::AccountId, amount: BalanceOf<T>) -> DispatchResult {
+		if !Self::can_hold(asset, who, amount) {
+			Err(TokenError::NoFunds)?
+		}
+		T::Store::mutate(
+			&(asset, who.clone()),
+			|extra| extra.reserved = extra.reserved.saturating_add(amount),
```
Comment (shawntabrizi) [inline frame/assets-freezer/src/lib.rs:240]: Why not checked add here?

File: frame/assets-freezer/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,361 @@
...
+		use fungibles::Transfer;
+		let result = Self::transfer(asset, source, dest, amount, WhenDust::KeepAlive);
+		if result.is_ok() {
+			if on_hold {
+				let r = Self::increase_on_hold(asset, dest, amount);
+				debug_assert!(r.is_ok(), "account exists and funds transferred in; qed");
+				r?
+			}
+		} else {
+			debug_assert!(false, "can_withdraw was successful; qed");
```
Comment (shawntabrizi) [inline frame/assets-freezer/src/lib.rs:295]: This is an assumption that user must always call `can_withdraw` before calling `transfer_held`?

Why can't we directly call it ourselves above?

File: frame/assets-freezer/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,361 @@
...
+	}
+
+	fn increase_on_hold(
+		asset: AssetIdOf<T>,
+		source: &T::AccountId,
+		amount: BalanceOf<T>,
+	) -> Result<(), DispatchError> {
+		T::Store::mutate(
+			&(asset, source.clone()),
+			|extra| extra.reserved = extra.reserved.saturating_add(amount)
```
Comment (shawntabrizi) [inline frame/assets-freezer/src/lib.rs:343]: could again be checked_add

File: frame/assets/src/lib.rs (hunk 2 above)
Comment (shawntabrizi) [reply frame/assets/src/lib.rs:426]: yeah i dont see why these extrinsics would ever need to be directly callable. The traits should provide access to all the functions you need.

File: frame/support/src/traits/tokens/misc.rs [hunk 10]
```diff
@@ -160,9 +205,9 @@ impl WithdrawReasons {
...
-pub trait AssetId: FullCodec + Copy + Default + Eq + PartialEq {}
-impl<T: FullCodec + Copy + Default + Eq + PartialEq> AssetId for T {}
+pub trait AssetId: FullCodec + Copy + Default + Eq + PartialEq + Debug {}
+impl<T: FullCodec + Copy + Default + Eq + PartialEq + Debug> AssetId for T {}
 
 /// Simple amalgamation trait to collect together properties for a Balance under one roof.
-pub trait Balance: AtLeast32BitUnsigned + FullCodec + Copy + Default {}
-impl<T: AtLeast32BitUnsigned + FullCodec + Copy + Default> Balance for T {}
+pub trait Balance: AtLeast32BitUnsigned + FullCodec + Copy + Default + Debug {}
+impl<T: AtLeast32BitUnsigned + FullCodec + Copy + Default + Debug> Balance for T {}
```
Comment (shawntabrizi) [inline frame/support/src/traits/tokens/misc.rs:213]: I think `RuntimeDebug` should be used?

File: frame/assets-freezer/src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,361 @@
...
+	T::Assets: fungibles::Transfer<T::AccountId>
+{
+	fn transfer(
+		asset: Self::AssetId,
+		source: &T::AccountId,
+		dest: &T::AccountId,
+		amount: Self::Balance,
+		death: WhenDust,
+	) -> Result<Self::Balance, DispatchError> {
+		T::Assets::transfer(asset, source, dest, amount, death)
```
Comment (shawntabrizi) [inline frame/assets-freezer/src/lib.rs:153]: Feeling a little stupid, but how is this right?

Shouldn't transfer check the freezer before delegating to the `Assets` implementation?

**Comment by shawntabrizi** [comment]
This is looking good to me at first glance.

File: frame/assets/src/lib.rs [hunk 12]
```diff
@@ -561,7 +565,8 @@ pub mod pallet {
 			let origin = ensure_signed(origin)?;
 			let who = T::Lookup::lookup(who)?;
 
-			let f = DebitFlags { keep_alive: false, best_effort: true };
+			let f = DebitFlags { keep_alive: false, ignore_freezer: false };
+			let amount = amount.min(Self::reducible_balance(id, &who, f)?);
 			let burned = Self::do_burn(id, &who, amount, Some(origin), f)?;
 			Self::deposit_event(Event::Burned(id, who, burned));
```
Comment (joepetrowski) [inline frame/assets/src/lib.rs:571]: Emitted in `do_burn`: https://github.com/paritytech/substrate/blob/efd262f/frame/assets/src/functions.rs#L345
```suggestion
			let _ = Self::do_burn(id, &who, amount, Some(origin), f)?;
```

File: frame/assets-freezer/src/lib.rs (hunk 8 above)
Comment (apopiak) [reply frame/assets-freezer/src/lib.rs:295]: It is called in `decrease_on_hold_ensuring_backed`

File: frame/support/src/traits/tokens/fungible/balanced.rs [hunk 13]
```diff
@@ -145,9 +155,12 @@ pub trait Balanced<AccountId>: Inspect<AccountId> {
 /// token imbalances in your system leading to accidental imflation or deflation. It's really just
 /// for the underlying datatype to implement so the user gets the much safer `Balanced` trait to
 /// use.
-pub trait Unbalanced<AccountId>: Inspect<AccountId> {
+pub trait Unbalanced<AccountId>: Inspect<AccountId> + Sized {
```
Comment (apopiak) [inline frame/support/src/traits/tokens/fungible/balanced.rs:158]: Where does the `Sized` come from?

File: frame/support/src/traits/tokens/misc.rs [hunk 14]
```diff
@@ -124,6 +124,50 @@ pub enum BalanceStatus {
...
+	/// resulting credit.
+	Credit,
+}
+
+impl WhenDust {
+	/// Return `true` if the account balance should be disposed of should it fall below minimum.
+	pub fn dispose(self) -> bool { matches!(self, WhenDust::Dispose) }
+
+	/// Return `true` if the account balance should neve be allowed to fall below minimum.
+	pub fn keep_alive(self) -> bool { matches!(self, WhenDust::KeepAlive) }
```
Comment (apopiak) [inline frame/support/src/traits/tokens/misc.rs:145]:
```suggestion
	/// Return `true` if the account balance should never be allowed to fall below minimum.
	pub fn keep_alive(self) -> bool { matches!(self, WhenDust::KeepAlive) }
	
	/// Return `true` if the account balance should be credited should it fall below minimum.
	pub fn credit(self) -> bool { matches!(self, WhenDust::Credit) }
```

File: frame/assets-freezer/src/lib.rs (hunk 7 above)
Comment (shawntabrizi) [reply frame/assets-freezer/src/lib.rs:240]: the point is we could fail here gracefully and without worry, so might as well do so.

**Comment by olanod** [comment]
No love for this one? is it blocked in some big way? just here to tell stale bot 🤖 that this is important and if there's anything us the simple mortals can do to help we'll be glad to.

**Comment by zauddelig** [comment]
what is blocking this issue?

**Comment by thiolliere** [comment]
AFAIK it is still relevant, it is just less priority than the other stuff

**Comment by zauddelig** [comment]
@thiolliere If I wanted to takeover what would be required to merge it?

**Comment by thiolliere** [comment]
> @thiolliere If I wanted to takeover what would be required to merge it?

When the PR was opened it only needed to be reviewed, now we also need to merge master. But it only makes sense to keep it up to date if people are available for reviewing it.

**Comment by zauddelig** [comment]
ok, so the PR is to be kept open, but no contribution is required?

**Pull request closed without merging**